sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOIT.java

```diff
@@ -196,11 +204,18 @@ public void testPublishingThenReadingAll() throws IOException {
     MetricsReader metricsReader = new MetricsReader(readResult, NAMESPACE);
     long actualRecords = metricsReader.getCounterMetric(READ_ELEMENT_METRIC_NAME);
 
+    // TODO(yathu) resolve pending messages with direct runner. Due to direct runner only finalize
+    //   checkpoint at very end, there are open consumers (may with buffer) and O(open_consumer)
+    //   message won't get delivered to other session
+    int remainRecords = countRemain(QUEUE);
+    assertTrue(remainRecords < OPTIONS.getNumberOfRecords() * 0.005);
+    LOG.info("has {} messages remain", remainRecords);
+
     assertTrue(
         String.format(
             "actual number of records %d smaller than expected: %d.",
-            actualRecords, OPTIONS.getNumberOfRecords()),
-        OPTIONS.getNumberOfRecords() <= actualRecords);
+            actualRecords, OPTIONS.getNumberOfRecords() - remainRecords),
+        OPTIONS.getNumberOfRecords() <= actualRecords + remainRecords);
```

Comment:
This assertion confused me. Are the total number of records `actualRecords` + `remainRecords`? To move this PR along, feel free to resolve this comment after answering.

Reply:
For actualRecords (records get read in pipeline), there are two category
- A: acknowledged records
- B: unacknowledged records that read

There is a third category, that is records not get read
- C: unacknowledged records that not read

remainRecords is unacknowledged message, and we have

remainRecords=B+C
OPTIONS.getNumberOfRecords() =A+B+C

that was why 
```
OPTIONS.getNumberOfRecords() <= actualRecords + remainRecords
```

I should rename "remainRecords" as "unackedRecords" for clarification

Reply:
Adjusted asserts to be more clear and added comments